ingest/src/main/java/com/microsoft/azure/kusto/ingest/ExponentialRetry.java

```diff
@@ -0,0 +1,61 @@
+package com.microsoft.azure.kusto.ingest;
+
+public class ExponentialRetry {
+    private final int maxAttempts;
+    private final double sleepBaseSecs;
+    private final double maxJitterSecs;
+    private int currentAttempt;
+
+    public ExponentialRetry(int maxAttempts) {
+        this(maxAttempts, 1.0, 1.0);
+    }
+
+    public ExponentialRetry(int maxAttempts, double sleepBaseSecs, double maxJitterSecs) {
+        this.maxAttempts = maxAttempts;
+        this.sleepBaseSecs = sleepBaseSecs;
+        this.maxJitterSecs = maxJitterSecs;
+        this.currentAttempt = 0;
+    }
+
+    public int getMaxAttempts() {
+        return maxAttempts;
+    }
+
+    public double getSleepBaseSecs() {
+        return sleepBaseSecs;
+    }
+
+    public double getMaxJitterSecs() {
+        return maxJitterSecs;
+    }
+
+    public double getCurrentSleepSecs() {
+        return sleepBaseSecs * (float) Math.pow(2, currentAttempt);
+    }
+
+    public int getCurrentAttempt() {
+        return currentAttempt;
+    }
+
+    public boolean shouldTry() {
+        return currentAttempt < maxAttempts;
+    }
+
+    public void doBackoff(){
+        if (!shouldTry()) {
+            throw new IllegalStateException("Max attempts exceeded");
```

Comment:
I would not expect throwing an exception when finding that we've already reached the maximum attempts.

Reply:
What is the correct way here?
The method is called while the object is in the wrong state

Reply:
I hear your point, but I wouldn't personally call it in the "wrong state". I'd say it's perfectly normal for `ExponentialRetry` to reach the max tries, but just its terminal state.

If we did want an exception, I'd expect a checked exception.

I'd have the method return a `boolean`.

Reply:
Wait, I'm looking now at how it's used (probably should've done that before), and see that you check `shouldRetry()` before calling `doBackoff()`, in which case if we've exhausted our tries we're indeed in an unexpected state.

How about simplifying it by removing `shouldRetry()` or at least making it `protected`, and `doBackoff()` would return false if it exhausted its retries. So the interface relevant to invoking this class would simply be `while(doBackoff())`.

Reply:
Like so?